src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1784,52 +1778,100 @@
+                  ;; them.  An external insertion of a fact that matches
+                  ;; a negation or accumulator condition can cause logical retractions.
+                  (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
+                          root alpha-roots]
+                    (alpha-activate root fact-group transient-memory transport transient-listener))
+                  (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))
+
+              :retraction
+              (do
+                (l/retract-facts! transient-listener facts)
+
+                (binding [*pending-external-retractions* (atom facts)]
+                  (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))))
+
+          (fire-rules* rulebase
+                       (:production-nodes rulebase)
+                       transient-memory
+                       transport
+                       transient-listener
+                       get-alphas-fn
+                       (uc/get-ordered-update-cache)))
+
+        #?(:cljs (throw (ex-info "Using the option for insertions and retractions to cancel each other out is not supported in ClojureScript"
+                                 {:session session :opts opts}))
+
+           :clj (do (let [insertions (into []
+                                           (comp (filter (fn [pending-op]
+                                                           (= (:type pending-op)
+                                                              :insertion)))
+                                                 (map :facts)
+                                                 cat)
+                                           pending-operations)
+                          retractions (into []
+                                            (comp (filter (fn [pending-op]
+                                                            (= (:type pending-op)
+                                                               :retraction)))
+                                                  (map :facts)
+                                                  cat)
+                                            pending-operations)
+                          update-cache (ca/get-cancelling-update-cache)]
+
+                      (binding [*current-session* {:rulebase rulebase
+                                                   :transient-memory transient-memory
+                                                   :transport transport
+                                                   :insertions (atom 0)
+                                                   :get-alphas-fn get-alphas-fn
+                                                   :pending-updates update-cache
+                                                   :listener transient-listener}]
+
+                        (doseq [[alpha-roots fact-group] (get-alphas-fn retractions)
```

Comment:
I'm still not clear on what your concerns are; could you give a more concrete example?  Obviously retracting will remove partial joins, but are you saying that you'd expect doing insertions first would end up doing less work overall?  Overall my goal here is to cut down on the work in the downstream Rete network, not eliminate the first level of work done in response to insertions and retractions.  Also, is this related to the optimization in AccumulateNode of delaying accumulation until there is a matching token?